analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs

```diff
@@ -141,59 +141,35 @@ ReportDiagnostic Severity(string id)
             }
         }
 
-        private static void CompareActualToExpected(string languageVersion, Diagnostic[] diagnostics, FileIssueLocations[] expectedIssuesPerFile, bool compareIdToMessage)
+        private static void CompareActualToExpected(Diagnostic[] diagnostics, CompilationIssues expected, bool compareIdToMessage)
         {
-            var actualIssues = new CompilationIssues(languageVersion, diagnostics);
-            actualIssues.Dump();
-
-            // ToDo: Throw this away
-            foreach (var diagnostic in diagnostics.OrderBy(x => x.Location.SourceSpan.Start))
+            var actual = new CompilationIssues(expected.LanguageVersion, diagnostics);
+            var pairs = MatchPairs(actual, expected);
+            var assertionMessages = new StringBuilder();
+            string previousFilePath = null;
+            foreach (var pair in pairs.OrderBy(x => x.FilePath).ThenBy(x => x.LineNumber).ThenBy(x => x.Start))
             {
-                var expectedIssues = ExpectedIssues(expectedIssuesPerFile, diagnostic.Location);
-                var issueId = VerifyPrimaryIssue(languageVersion,
-                    expectedIssues,
-                    issue => issue.IsPrimary,
-                    diagnostic.Location,
-                    compareIdToMessage ? diagnostic.Id : diagnostic.GetMessage(),
-                    compareIdToMessage
-                        ? $"{languageVersion}: Unexpected build error [{diagnostic.Id}]: {diagnostic.GetMessage()} on line {diagnostic.Location.GetLineNumberToReport()}"
-                        : null);
-
-                var secondaryLocations = diagnostic.AdditionalLocations
-                    .Select((_, i) => diagnostic.GetSecondaryLocation(i))
-                    .OrderBy(x => x.Location.GetLineNumberToReport())
-                    .ThenBy(x => x.Location.GetLineSpan().StartLinePosition.Character);
-
-                foreach (var secondaryLocation in secondaryLocations)
+                if (previousFilePath != pair.FilePath)
```

Comment:
this `previousFilePath` variable seems like a roundabout way of doing a groupBy. 
What about something like this:

```csharp
foreach (var pairsByFile in pairs.GroupBy(x => x.FilePath))
{
    if (assertionMessages.Length > 0)
    {
        assertionMessages.AppendLine();
    }
    assertionMessages.AppendLine($"There are differences for {actual.LanguageVersion} {pairsByFile.Key}:");
    foreach (var pair in pairsByFile.OrderBy(x => x.LineNumber).ThenBy(x => x.Start))
    {
        pair.AppendAssertionMessage(assertionMessages);
    }
}
```

IMO it reads a bit more naturally, since you don't have to mentally track the state of the `previousFilePath`.